### ISSUE TITLE
Using the Bitbucket provider

### ISSUE THREAD

**Issue by ryankanno** [jonhadfield/soba#5]
Great tool! I just wanted to leave a note for peeps who may use this for Bitbucket.

Needed to do two things to make this work after creating the OAuth consumer.

1. Authorize the OAuth consumer to my account by visiting https://bitbucket.org/site/oauth2/authorize?client_id=<client_id>&response_type=code once - fill in your client_id from the OAuth consumer creation step.
2. Set the BITBUCKET_USER env variable to ~~x-auth-token~~ x-token-auth as described [here](https://support.atlassian.com/bitbucket-cloud/docs/use-oauth-on-bitbucket-cloud/#Cloning-a-repository-with-an-access-token).

Curious if your BITBUCKET_USER is set to your username because for some reason, I kept getting git 128 access errors with it.

**Comment by jonhadfield** [comment]
Hi @ryankanno, thanks for the feedback.
In my test setup and I'm using my username in BITBUCKET_USER. Is there any more detail with that error message?

**Comment by ryankanno** [comment]
Interesting. I'm running the following:

```
docker run --rm -t -v bitbucket-backup:/backup \
    -e GIT_BACKUP_DIR=/backup \
    -e BITBUCKET_USER=*** \
    -e BITBUCKET_KEY=xxx \
    -e BITBUCKET_SECRET=yyy \
    quay.io/jonhadfield/soba
```
where xxx,yyy are from the OAuth consumer creation step.

If I leave BITBUCKET_USER (***) as my username, I get the following:

![Screen Shot 2021-08-24 at 4 55 23 AM](https://user-images.githubusercontent.com/57014/130511222-106ff97c-a54f-4a07-b58f-9479122c2df6.png)

If I switch it to x-token-auth as I describe above, I get the following:

![Screen Shot 2021-08-24 at 5 03 56 AM](https://user-images.githubusercontent.com/57014/130512139-ee237ee2-ed34-4dda-91bf-bbfd27ac3c3f.png)

**Comment by ryankanno** [comment]
I just wanted to update this ticket because I wanted to start 'fresh' to see what the actual issue is. In my [original comment](https://github.com/jonhadfield/soba/issues/5#issue-976311622), I mentioned that I needed to do the following:

1. Authorize the OAuth consumer to my account by visiting https://bitbucket.org/site/oauth2/authorize?client_id=<client_id>&response_type=code once - fill in your client_id from the OAuth consumer creation step.

I was incorrect, and this is an unnecessary step. I looked more closely at what the underlying [githosts-utils library](https://github.com/jonhadfield/githosts-utils/blob/master/bitbucket.go#L19) was doing, and it's using the client_credentials grant. To enable this, you do need to select the *This is a private consumer* option when creating the OAuth consumer. 

With that said, I still do need to set the user to x-token-auth instead of my actual username. 

One thing to note is that even with the client_credentials grant, I only was able to access the public/private repositories in the workspace where I created the OAuth consumer. I couldn't find anything in the BitBucket docs that indicated if that was by design - even when I opened up the scopes of the OAuth consumer. I can get around this by just creating consumers in each workspace, but one idea may be to support app passwords instead of the key/secret dance which would simplify a bit of this.

Thanks for all your work on this!
